packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -1943,6 +1944,51 @@ export class MergeTree {
         }
     }
 
+    /**
+     * Perform local rollback of previously deleted segments
+     */
+    public rollbackRemovedSegment(segments: ISegment[]) {
+        for (const segment of segments) {
+            segment.removedClientIds = undefined;
+            segment.removedSeq = undefined;
+            segment.localRemovedSeq = undefined;
+
+            if (this.mergeTreeDeltaCallback) {
+                const insertOp = createInsertSegmentOp(this.findRollbackPosition(segment), segment);
+                this.mergeTreeDeltaCallback(
+                    { op: insertOp },
+                    {
+                        operation: MergeTreeDeltaType.INSERT,
+                        deltaSegments: [{ segment }],
+                    });
+            }
+        }
+
+        this.blockUpdate(this.root);
+    }
```

Comment:
I think the local references that slid when the segments were removed should move back. I'm not sure how that should be done. Maybe putting the original info into the localOpMetadata?

Reply:
Yeah, putting that info in localOpMetadata seems good to me. Note you need to do this for both slid references and detached references (should be everything except StayOnRemove ones).

Reply:
The slid references only seem to get completed when the remove is ack'ed which doesn't happen within orderSequentially and rollback. Let me know if that's not right or I should make it clearer in the code.

Reply:
oops, I was thinking about the Stay -> SlideOnRemove transition that intervalCollection does when that isn't really relevant. Yeah, it's correct that even SlideOnRemove references don't slide until the removal of a segment is acked.